stream_alert/shared/config.py

```diff
@@ -0,0 +1,154 @@
+"""
+Copyright 2017-present, Airbnb Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+from collections import defaultdict, OrderedDict
+import json
+import os
+
+
+class ConfigError(Exception):
+    """Exception class for config file errors"""
+
+
+def parse_lambda_arn(function_arn):
+    """Extract info on the current environment from the lambda function ARN
+
+    Parses the invoked_function_arn from the given context object to get
+    the name of the currently running alias (either production or development)
+    and the name of the function.
+
+    Example:
+        arn:aws:lambda:aws-region:acct-id:function:stream_alert:production
+
+    Arguments:
+        function_arn (str): The AWS Lambda function ARN
+
+    Returns:
+        dict:
+            {
+                'region': 'region_name',
+                'account_id': 'account_id',
+                'function_name': 'function_name',
+                'qualifier': 'qualifier'
+            }
+    """
+    split_arn = function_arn.split(':')
+    return {
+        'region': split_arn[3],
+        'account_id': split_arn[4],
+        'function_name': split_arn[6],
+        'qualifier': split_arn[7] if len(split_arn) == 8 else None # optional qualifier
+    }
+
+
+def load_config(conf_dir='conf/', **kwargs):
+    """Load the configuration for StreamAlert.
+
+    All configuration files live in the `conf` directory in JSON format.
+    `sources` define a colleciton of AWS services (S3, Kinesis) supported as
+    inputs to StreamAlert, specific entities (S3 buckets, Kinesis streams),
+    and log types emitted from them.
+
+    `logs` declare the schema for the listed log types in `sources`.  Each
+    key denotes the name of the log type, and includes 'keys' used to match
+    rules to log fields.
+
+    Arguments:
+        conf_dir (str): [optional] Path from which to load the config
+
+    Keyword Arguemnts:
+        exclude (set): Names of config files or folders that should not be loaded
+        include (set): Names of specific config files to only load
```

Comment:
Out of curiosity, what was the motivation for selectively loading parts of the config? Performance?

A potential downside to this approach is having to reason about which sections of the config are available to different Lambda functions instead of being able to reason about a single config object.

A potential future iteration of this (once we have a Config class) would be to lazy-load attributes. That way, callers don't need to explicitly say which config they need

Reply:
I suppose the motivation was largely to allow for supporting what we _currently_ do. For instance, the alert processor only loads the `outputs.json`, so I thought supporting a similar model was ideal. While I agree it could be confusing to have to reason about which files are loaded and which are not, the caller is the one who _restricts_ the loaded files (and thus should be responsible for knowing this..) otherwise all of them will be loaded.

After this initial iteration I'm open to building it out more (and anticipate doing so). I like the idea of lazy loading and think that we can benefit a lot from that approach!